[CIR] Upstream CastOp and scalar conversions

mmha opened the pull request:
This patch upstreams ClangIR's CastOp with the following exceptions:
- No Fixed/FP conversions
- No casts between value categories
- No complex casts
- No array_to_ptrdecay
- No address_space
- No casts involving record types (member pointers, base/derived casts)
- No casts specific to ObjC or OpenCL

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        !cgf.getContext().getLangOpts().NativeHalfType) {
+      // Cast to FP using the intrinsic if the half type itself isn't supported.
+      if (mlir::isa<cir::CIRFPTypeInterface>(dstTy)) {
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics())
```
Is the fallthrough of this to the srcType/srcTy assignmetn below acceptable?

andykaylor replied:
I think we could just issue an error if `useFP16ConversionIntrinsics()` is true but create the cast as if it were not. So the code as written here won't work because it wouldn't get to the `createCast` call.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+      // Cast to FP using the intrinsic if the half type itself isn't supported.
+      if (mlir::isa<cir::CIRFPTypeInterface>(dstTy)) {
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+          cgf.getCIRGenModule().errorNYI(loc,
```
These don't terminate the function, all errorNYI has to SOMEHOW graciously recover from this function, so perhaps we should return SOMETHING here?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        // FPExt, depending on whether the half type itself is supported (as
+        // opposed to operations on half, available with NativeHalfType).
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
+          cgf.getCIRGenModule().errorNYI(loc,
```
Same here.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // other pointers and integers. Check for pointer types in terms of LLVM, as
+    // some native types (like Obj-C id) may map to a pointer type.
+    if (auto dstPT = dyn_cast<cir::PointerType>(dstTy)) {
+      cgf.getCIRGenModule().errorNYI(loc, "pointer casts");
```
this feels like it should definitely return something other than falling through to below.

erichkeane replied:
If this falls through it could very well result in a pointer-to-pointer cast being done as pointer-to-int, which is likely to create nonsense and likely cause crashes later.  We should probably just create a constant of `dest` type here.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    }
+
+    if (srcType->isMatrixType() && dstType->isMatrixType())
+      cgf.getCIRGenModule().errorNYI(loc,
```
here too.  it ends up just falling through to the assert below which is awkward.

erichkeane replied:
still valid here, I think we need to just generate a constant of the dest type.

erichkeane replied:
perhaps a `return nullptr' here?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -130,17 +389,136 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *ce) {
 
+  // These cases are generally not written to ignore the result of evaluating
+  // their sub-expressions, so we clear this now.
+  [[maybe_unused]] bool ignored = TestAndClearIgnoreResultAssign();
```
```suggestion
  TestAndClearIgnoreResultAssign();
```

More interestingly, why not just ignoreResultAssign = false; or something?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -130,17 +389,136 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *ce) {
+    break;
+  case CK_NonAtomicToAtomic:
+  case CK_UserDefinedConversion:
+    return Visit(const_cast<Expr *>(e));
```
would VERY much like a rename of `e` to `subExpr` or something.  This got VERY confusing to review.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -130,17 +389,136 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *ce) {
+      // CK_NoOp can model a pointer qualification conversion, which can remove
+      // an array bound and change the IR type.
+      // FIXME: Once pointee types are removed from IR, remove this.
+      auto t = convertType(destTy);
```
not a place we can use `auto`.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -130,17 +389,136 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *ce) {
+  case CK_BlockPointerToObjCPointerCast:
+  case CK_AnyPointerToBlockPointerCast:
+  case CK_BitCast: {
+    auto src = Visit(const_cast<Expr *>(e));
```
another place we can't use auto.

erichkeane reviewed (commented):
We have to be mroe careful I think with uses of the `errorNYI` in a few places, we are using it too much like `assert` when it is more like `emitDiag`. 

Also, this patch uses `auto` a bunch in ways that are contrary to the coding standard.  See : https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

`auto` is really only to be used (rule of thumb) when the type is already 'listed' on the RHS, or naming the type is so obnoxious as to be irrelevant (iterators/etc).

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -36,6 +36,18 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+  //                               Utilities
+  //===--------------------------------------------------------------------===//
+
+  bool TestAndClearIgnoreResultAssign() {
```
I noticed that the underlying variable is never read from, making this dead code. I removed it.

mmha replied:
It's dead code in ClangIR and in the incubator, but classic codegen reads this when visiting complex expressions. I put it back in as a simple `ignoreResultAssign = false` because it's probably harder to backtrack all the places where we need to set this when we add complex expression than to just do it now.

mmha committed:
Update clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

Co-authored-by: Erich Keane <ekeane@nvidia.com>

andykaylor commented on clang/test/CIR/CodeGen/cast.cpp:
```diff
@@ -0,0 +1,58 @@
+  // CHECK: %{{[0-9]+}} = cir.cast(integral, %{{[0-9]+}} : !cir.int<u, 32>), !cir.int<s, 32>
+
+  unsigned uu = (unsigned)x1; // should not be generated
+  // CHECK-NOT: %{{[0-9]+}} = cir.cast(integral, %{{[0-9]+}} : !cir.int<u, 32>), !cir.int<u, 32>
```
I think this check is too specific for CHECK-NOT. Maybe move the cases that shouldn't cast to a separate function and just CHECK-NOT for `cir.cast` between the function start and end braces.

andykaylor commented on clang/test/CIR/CodeGen/cast.cpp:
```diff
@@ -0,0 +1,58 @@
+// RUN: %clang_cc1 -std=c++17 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
```
Can you expand this test to cover all the cast types you're adding (or remove any that can't be supported yet)?

bcardosolopes replied:
Has this been addressed?

mmha replied:
I removed all tests for the currently unsupported casting kinds. I noticed that a few were missing (bool to int/float/ptr) which I readded.

andykaylor replied:
It looks like floating casts are still missing. Can you add float-to-double and double-to-float test?

mmha commented on clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h:
```diff
@@ -78,6 +79,67 @@ class CIRBaseBuilderTy : public mlir::OpBuilder {
+    return createBitcast(src, getPointerTo(newPointeeTy));
+  }
+
+  mlir::Value createAddrSpaceCast(mlir::Location loc, mlir::Value src,
```
Correct. I removed these two functions again for now.

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -84,26 +96,266 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
       return src;
+    }
+
+    if (dstType->isVoidType())
```
Do you have an example of having this info retained in CIR is useful?

erichkeane replied:
The explicit ignoring of a value shows a level of intent that just ignoring the return value does not.  Whether this permits some sort of intuition to happen as an optimization opportunity is perhaps beyond my level of creativity, but one that I still wonder if could have some value in CIR.  This is more of a `we should think about this` than a `we should change something` here.

mmha committed:
Address some review feedback

- Remove dead functions
- Remove convertType helper function
- rename some variables
- change __fp16 conversion codegen to emit cast even if intrinsic is requested
- Move some CHECK-NOT tests out into a separate function

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,20 +89,259 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+                             mlir::Type dstTy, ScalarConversionOpts opts) {
+    assert(!srcType->isMatrixType() && !dstType->isMatrixType() &&
+           "Internal error: matrix types not handled by this function.");
+    if (mlir::isa<mlir::IntegerType>(srcTy) ||
```
I realize this made sense in the case of 'switching from this function to another' in the incubator, but I think this should probably be an assert instead.

erichkeane replied:
Same comment again.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,20 +89,259 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // conversions to fixed point types from other types.
+    // conversions to fixed point types from other types.
+    if (srcType->isFixedPointType())
+      cgf.getCIRGenModule().errorNYI(loc, "fixed point conversions");
```
These also need to return something.  Otherwise this would fall into the below and it isn't clear it would fall into anything even vaguely sensible.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,20 +89,259 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // Handle pointer conversions next: pointers can only be converted to/from
+    // other pointers and integers. Check for pointer types in terms of LLVM, as
+    // some native types (like Obj-C id) may map to a pointer type.
+    if (auto dstPT = dyn_cast<cir::PointerType>(mlirDstType)) {
```
```suggestion
    if (isa_cast<cir::PointerType>(mlirDstType)) {
```

else this is a not-used warning.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,20 +89,259 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+                 srcType.getTypePtr() &&
+             "Splatted expr doesn't match with vector element type?");
+
+      llvm_unreachable("not implemented");
```
Any reason this is a `unreachable` instead of a NYI?  Also, `unreachable` isnt' really the tool for `not yet implemented`, we typically want to use `assert` for reasons that the lead maintainer explained to me in the past but I've failed to internalize.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -121,29 +359,159 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
+  case CK_AtomicToNonAtomic:
+    cgf.getCIRGenModule().errorNYI(subExpr->getSourceRange(),
+                                   "CastExpr: ", ce->getCastKindName());
+    break;
```
what does this fall through to? Does it result in a sensible value?

erichkeane replied:
still, atomic-to-non-atomic cast can change types, so probably should return something.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -121,29 +359,159 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
+  case CK_PointerToIntegral: {
+    assert(!destTy->isBooleanType() && "bool should use PointerToBool");
+    if (cgf.cgm.getCodeGenOpts().StrictVTablePointers)
+      llvm_unreachable("NYI");
```
see above about unreachable/assert/NYI.

mmha commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -78,6 +78,111 @@ class LLVMLoweringInfo {
+// CastOp
+//===----------------------------------------------------------------------===//
+
+// The enumaration value isn't in sync with clang.
```
I tried to get the enums in sync. Good news is I found a couple of unused casting kinds wrt. complex numbers which I removed. Bad news is I found one casting kind (bool to float) that's specific to CIR: llvm/clangir#290. I moved it to the bottom on the list so the enum values of classic CG and CIR match up.

mmha committed:
More review feedback changes

- CastKind enum now aligns with classic codegen CastKind
- Try to emit a value of the expected type after errorNYI()s

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -78,6 +78,156 @@ class LLVMLoweringInfo {
+// CastOp
+//===----------------------------------------------------------------------===//
+
+// CK_Dependent
```
Neat!  Thank you!

erichkeane commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -247,6 +280,179 @@ struct ConvertCIRToLLVMPass
+  switch (castOp.getKind()) {
+  case cir::CastKind::array_to_ptrdecay: {
+    const auto ptrTy = mlir::cast<cir::PointerType>(castOp.getType());
+    auto sourceValue = adaptor.getOperands().front();
```
A LOT of uses of 'auto' in this function that don't match our coding standard.

mmha committed:
Update clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

Co-authored-by: Erich Keane <ekeane@nvidia.com>

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+  }
+
+  mlir::Value emitFloatToBoolConversion(mlir::Value src, mlir::Location loc) {
+    auto boolTy = builder.getBoolTy();
```
probably can't do `auto` here either.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // as a logical value again.
+    // TODO: optimize this common case here or leave it for later
+    // CIR passes?
+    mlir::Type boolTy = cgf.convertType(cgf.getContext().BoolTy);
```
why convertType here but `getBoolTy` on 106?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+
+    if (llvm::isa<MemberPointerType>(srcType)) {
+      cgf.getCIRGenModule().errorNYI(loc, "member pointer to bool conversion");
+      auto boolType = cgf.getContext().getBOOLType();
```
Another case we can't use autos.  Also a 3rd way to get a `bool` type?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+      auto addr =
+          builder.createAlloca(loc, builder.getPointerTo(cirBoolType),
+                               cirBoolType, {}, cgf.cgm.getSize(alignment));
+      return builder.createLoad(loc, addr);
```
I think you mentioned we dont have constants implemented?  Else just a bool constant would be sufficient I would expect.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+      auto addr =
+          builder.createAlloca(src.getLoc(), builder.getPointerTo(dstPT), dstPT,
+                               {}, cgf.cgm.getSize(alignment));
+      return builder.createLoad(src.getLoc(), addr);
```
Wonder if there is use of a `createDummyValue` function in builder that does all this for us?

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    res = emitScalarCast(src, srcType, dstType, mlirSrcType, mlirDstType, opts);
+
+    if (mlirDstType != resTy) {
+      if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
```
no curleys on single-line if.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -121,29 +375,174 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
 
+  // These cases are generally not written to ignore the result of evaluating
+  // their sub-expressions, so we clear this now.
+  ignoreResultAssign = false;
```
Is this variable used?

mmha replied:
See my comment here: https://github.com/llvm/llvm-project/pull/130690#discussion_r1990214545

At the moment no, but I think it would be more error prone to add it later than to add it as a dead variable now.

andykaylor replied:
I agree.

erichkeane replied:
You'll have to do a void cast on it then, else this will cause a warning.

erichkeane replied:
I spent a while trying to repro the warning I could swear I've seen before, but I've not been able to get this to happen.  So disregard.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -121,29 +375,174 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
+    if (kind == CK_FixedPointToFloating || kind == CK_FloatingToFixedPoint) {
+      cgf.getCIRGenModule().errorNYI(subExpr->getSourceRange(),
+                                     "fixed point casts");
+      return {};
```
what is the difference between `return nullptr` and `return {}` for this type?

andykaylor replied:
There should be no difference. mlir::Value is a wrapper around an implementation pointer, so returning `nullptr` will construct an mlir::Value with a null implementation, and the default constructor does the same. Of course, that could change. I tried to standardize on `{}` in the constant emitter.

mmha replied:
Good to know you standardized on `{}`. I changed all instances of `return nullptr` to that.

mmha committed:
Addressing more review feedback
- Add createDummyValue functions
- Create Bool/Ptr constants on errorNYIs
- return {} instead of nullptr
- Replace auto with types

mmha committed:
Add support for CK_ToVoid

mmha committed:
clang-format

erichkeane commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -247,6 +280,179 @@ struct ConvertCIRToLLVMPass
+    mlir::Type targetType = convertTy(ptrTy);
+    mlir::Type elementTy = convertTypeForMemory(*getTypeConverter(), dataLayout,
+                                                ptrTy.getPointee());
+    auto offset = llvm::SmallVector<mlir::LLVM::GEPArg>{0};
```
```suggestion
    llvm::SmallVector<mlir::LLVM::GEPArg> offset{0};
```

erichkeane reviewed (approved):
1 nit, else LGTM.

mmha committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Erich Keane <ekeane@nvidia.com>

bcardosolopes reviewed (approved):
LGTM pending on question left in one comment

mmha committed:
- Add more test cases (int_to_bool, ptr_to_bool, float_to_bool)
- Handle IgnoredExpr on null_to_pointer casts

mmha committed:
fix test

mmha committed:
clang-format

andykaylor commented on clang/test/CIR/CodeGen/cast.cpp:
```diff
@@ -0,0 +1,79 @@
+// CHECK: cir.func @cStyleCasts_0
+
+  char a = (char)x1; // truncate
+  // CHECK: %{{[0-9]+}} = cir.cast(integral, %{{[0-9]+}} : !cir.int<u, 32>), !cir.int<s, 8>
```
Maybe add checks for lowering (to LLVM IR through clangir and using classic codegen) here also?

mmha replied:
I expanded the tests to include LLVM lowering where feasible. `int_to_bool` is implemented in CIR but not in the lowering (as that requires comparisons) so I `ifdef`ed those out for the LLVM checks.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -121,29 +364,173 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
 
+  // These cases are generally not written to ignore the result of evaluating
+  // their sub-expressions, so we clear this now.
+  ignoreResultAssign = false;
```
The incubator does use the value returned by `TestAndClearIgnoreResultAssign()` in `VisitBinAssign()` and `emitCompoundAssign()` so we'll probably want to keep the call in those places, but I don't see any value in calling the function when we aren't using the value returned.

mmha replied:
Note that `ignoreResultAssign` is a member variable and retains its value while we visit the expression. We might get called recursively so we should keep this store.

andykaylor replied:
Yes, I agree that keeping the store is necessary. My point was that even in the incubator this value is being used.

mmha committed:
- Correctly emit dummy value for floating casts
- Emit dummy value and a diagnostic during LLVM lowering for int_to_bool casts
- Expand tests to cover LLVM lowering

andykaylor commented on clang/test/CIR/CodeGen/cast.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+// CIR: cir.func @cxxstaticcast_0
+// CIR:    %0 = cir.alloca !cir.int<u, 32>, !cir.ptr<!cir.int<u, 32>>, ["x", init] {alignment = 4 : i64}
```
You should use pattern matching for %0 and similar values.

andykaylor reviewed (approved):
LGTM with one very minor nit

mmha committed:
Use pattern matching for CIR/LLVM registers

mmha committed:
[CIR] Upstream CastOp and scalar conversions

This patch upstreams ClangIR's CastOp with the following exceptions:
- No Fixed/FP conversions
- No casts between value categories
- No complex casts
- No array_to_ptrdecay
- No address_space
- No casts involving record types (member pointers, base/derived casts)
- No casts specific to ObjC or OpenCL

mmha committed:
Fix MissingFeatures.h
